lib/liquid/static_registers.rb

```diff
@@ -25,8 +25,12 @@ def delete(key)
       @registers.delete(key)
     end
 
-    def fetch(key, default = nil)
-      key?(key) ? self[key] : default
+    def fetch(key, *args, &block)
+      if @registers.key?(key)
+        @registers.fetch(key)
+      else
+        @static.fetch(key, *args, &block)
```

Comment:
Using `*args` parameter will end up unnecessarily allocating an array.  So this could cause a performance problem if it is called often enough.  Although, it doesn't look like we are using this method internally to liquid, so it might not matter.

Reply:
This would be whenever filters try to access registries. It could indeed be called multiple times depending on the use in filters.

I'm open to suggestions if any. 

If we really wanted to go after performance, we could consider generating a single unified Hash whenever there is a mutation and avoid checkout 2 of them every call.

Reply:
If we want to avoid the performance overhead, then you could use

```suggestion
    UNDEFINED = Object.new

    def fetch(key, default=UNDEFINED, &block)
      if @registers.key?(key)
        @registers.fetch(key)
      elsif default != UNDEFINED
        @static.fetch(key, default, &block)
      else
        @static.fetch(key, &block)
```

Reply:
> If we really wanted to go after performance, we could consider generating a single unified Hash whenever there is a mutation and avoid checkout 2 of them every call.

Yeah, this might make more sense.  If we typically mutate the registers, we could just unconditionally merge them ahead of time, which would be quite simple